bin/cron/hoc_student_name_cleanup

```diff
@@ -0,0 +1,224 @@
+  ) do |batch_size|
+    options[:batch_size] = batch_size
+  end
+
+  opts.on('--sleep=0',
+    Integer,
+    'Amount of time in milliseconds to sleep between batches. Higher values ' \
+    'will reduce the performance impact on the datbase. Defaults to 0 msecs ' \
+    'if not specified.'
+  ) do |sleep_msecs|
+    options[:sleep_msecs] = sleep_msecs
+  end
+
+  opts.on('--dry-run',
+    'Enables read-only mode where no changes are written to the database'
+  ) do |dry_run|
+    options[:dry_run] = dry_run
+  end
+
+  opts.on('--log-sql',
+    'Logs all sql statements sent to the database'
+  ) do |log_sql|
+    options[:log_sql] = log_sql
+  end
+
+  opts.on('--[no-]slack',
+    'Sends summary of script to slack. Defaults to true.'
+  ) do |notify_slack|
+    options[:notify_slack] = notify_slack
+  end
+
+  opts.on('-h', '--help', 'Prints this help message') do
+    puts opts
+    exit
+  end
+end.parse!
+CDO.log.info "Initial parsed options: #{options}"
+
+PEGASUS_DB.loggers << CDO.log if options[:log_sql]
+
+if options[:start_date].nil?
+  last_processed_date = Properties.get(LAST_PROCESSED_DATE_PROPERTY)
+  if last_processed_date.nil?
+    CDO.log.error "--start-date is required on first run of the script"
+    return
+  end
+  options[:start_date] = Date.parse(last_processed_date) + 1
+end
+
+if options[:end_date].nil?
+  options[:end_date] = Date.today - options[:retention_days]
+end
+
+# Options should not change from here on
+options.freeze
+CDO.log.info "Final calculated options: #{options}"
+
+# Main loop - Conceptually, we want to remove names from all rows with a
+# finished_at time between the start date and end date. We iterate over the
+# rows with a set of nested loops. The outer loop processes one day at a time,
+# checkpointing after each iteration by updating LAST_PROCESSED_DATE_PROPERTY
+# after each iteration. The inner loop processes the rows for a given day in
+# small batches to minimize impact to the database.
+date = options[:start_date]
+total_rows_processed = 0
+total_rows_updated = 0
+while date < options[:end_date]
+  CDO.log.info "Processing rows with finished_at=#{date}"
+
+  batch_start_time = date.to_time
+  batch_start_id = 0
+  batch_end_time = (date + 1).to_time
+  batch_number = 0
+  batch_rows_updated = 0
+
+  while batch_start_time < batch_end_time
+    ids_to_cleanup = []
+    last_row = nil
+
+    # Construct the sql where clauses for the start and end of the interval for
+    # this batch. Rows are ordered by (finished_at, id). The interval starts at
+    # (batch_start_time, batch_start_id) inclusive and ends at (batch_end_time, 0)
+    # exclusive.
+    interval_start_where_clause = Sequel.lit(
+      '(finished_at = ? AND id >= ?) OR finished_at > ?',
+      batch_start_time, batch_start_id, batch_start_time
+    )
+    interval_end_where_clause = Sequel.lit(
+      'finished_at < ?',
+      batch_end_time
+    )
+
+    HOC_ACTIVITY.select(:id, :name, :finished_at).
```

Comment:
Can we assess the performance of these queries by running it on a clone of the production database?

Reply:
Yes, I've done a lot of experimenting and testing on the single query while developing this script and verified that this query uses available indexes and and returns quickly.  I will assess performance when running the whole script in the final verification pass (like we did for the initial cleanup script).